### PR TITLE
Add migration schedules trigger to stork async DR longevity

### PR DESCRIPTION
**Pull request by gejain**
**What this PR does / why we need it**:
"Add migration schedules trigger to stork async DR longevity"

**Which issue(s) this PR fixes** (optional)
Closes #

**Special notes for your reviewer**:
Successful Run - https://aetos.pwx.purestorage.com/resultSet/testSetID/210271/testCaseID/1105609

### CODE REVIEW

File: pkg/asyncdr/asyncdr.go [hunk 1]
```diff
@@ -97,6 +97,119 @@ func CreateMigration(
...
+			Namespace: namespace,
+		},
+		Spec: storkapi.MigrationScheduleSpec{
+			Template:           TemplateSpec,
+			SchedulePolicyName: sp,
+			Suspend:            suspend,
+			AutoSuspend:        autoSuspend,
+		},
+	}
+	mig_sched, err := storkops.Instance().CreateMigrationSchedule(migrationSchedule)
```
Comment (Rohit-PX) [inline pkg/asyncdr/asyncdr.go:136]:
```suggestion
	migSched, err := storkops.Instance().CreateMigrationSchedule(migrationSchedule)
```

Comment (Rohit-PX) [inline pkg/asyncdr/asyncdr.go:136]:
```suggestion
	migSched, err := storkops.Instance().CreateMigrationSchedule(migrationSchedule)
```

File: pkg/asyncdr/asyncdr.go [hunk 2]
```diff
@@ -97,6 +97,119 @@ func CreateMigration(
...
+	name string,
+	namespace string,
+	clusterPair string,
+	migrationNamespace string,
+	includeVolumes *bool,
+	includeResources *bool,
+	startApplications *bool,
+	sp string,
+	suspend *bool,
+	autoSuspend bool,
```
Comment (Rohit-PX) [inline pkg/asyncdr/asyncdr.go:110]: Need to add all the parameters for migration schedule: https://github.com/libopenstorage/stork/blob/7ee546008cb1c09acf7d0cb9ad0d1a223486cd39/pkg/apis/stork/v1alpha1/migration.go#L15

That way we can create all types of migration schedules possible, if required.

File: pkg/asyncdr/asyncdr.go [hunk 3]
```diff
@@ -97,6 +97,119 @@ func CreateMigration(
...
+				},
+			}}
+		schedPolicy, err = storkops.Instance().CreateSchedulePolicy(schedPolicy)
+	} else {
+		log.Infof("schedPolicy is %v already exists", schedPolicy.Name)
+	}
+	return schedPolicy, err
+}
+
+func WaitForNuOfMigration(sched_name string, sched_namespace string, count int, miginterval int) (map[string]string, error) {
```
Comment (Rohit-PX) [inline pkg/asyncdr/asyncdr.go:160]: Add comments about what the function is doing.
Eg. 
```
      //WaitForNumOfMigration waits for a certain number of migrations to complete.

Comment (Rohit-PX) [inline pkg/asyncdr/asyncdr.go:160]:
```suggestion
func WaitForNumOfMigration(sched_name string, sched_namespace string, count int, miginterval int) (map[string]string, error) {
```

File: pkg/asyncdr/asyncdr.go [hunk 4]
```diff
@@ -97,6 +97,119 @@ func CreateMigration(
...
+	migInterval := time.Minute * time.Duration(miginterval)
+	migTimeout := time.Minute * time.Duration(count*miginterval)
```
Comment (Rohit-PX) [inline pkg/asyncdr/asyncdr.go:162]: The total number of migrations i.e `count` should not affect the time taken per migration. So this calculation would not be an accurate estimate of timeout per migration.

File: pkg/asyncdr/asyncdr.go [hunk 5]
```diff
@@ -97,6 +97,119 @@ func CreateMigration(
...
+	expected_migrations := make(map[string]string)
```
Comment (Rohit-PX) [inline pkg/asyncdr/asyncdr.go:163]:
```suggestion
	expectedMigrations := make(map[string]string)
```

File: pkg/asyncdr/asyncdr.go [hunk 6]
```diff
@@ -97,6 +97,119 @@ func CreateMigration(
...
+	checkNuOfMigrations := func() (interface{}, bool, error) {
```
Comment (Rohit-PX) [inline pkg/asyncdr/asyncdr.go:164]:
```suggestion
	checkNumOfMigrations := func() (interface{}, bool, error) {
```

File: pkg/asyncdr/asyncdr.go [hunk 7]
```diff
@@ -97,6 +97,119 @@ func CreateMigration(
...
+		response, err := storkops.Instance().GetMigrationSchedule(sched_name, sched_namespace)
+		if err != nil {
+			return "", true, fmt.Errorf("failed to get migrationschedule: %v", sched_name)
+		}
+		migs := response.Status.Items["Interval"]
+		for _, mig := range migs {
+			var migrations []*storkapi.Migration
+			if _, ok := expected_migrations[mig.Name]; !ok {
```
Comment (Rohit-PX) [inline pkg/asyncdr/asyncdr.go:172]:
```suggestion
			if _, ok := expectedMigrations[mig.Name]; !ok {
```

File: pkg/asyncdr/asyncdr.go [hunk 8]
```diff
@@ -97,6 +97,119 @@ func CreateMigration(
...
+				migration, err := storkops.Instance().GetMigration(mig.Name, sched_namespace)
+				if err != nil {
+					return "", true, fmt.Errorf("failed to get migration: %v", mig.Name)
+				}
+				migrations = append(migrations, migration)
+				err = WaitForMigration(migrations)
+				if err != nil {
+					expected_migrations[mig.Name] = fmt.Sprintf("Migration failed with error: %v", err)
```
Comment (Rohit-PX) [inline pkg/asyncdr/asyncdr.go:180]:
```suggestion
					expectedMigrations[mig.Name] = fmt.Sprintf("Migration failed with error: %v", err)
```

File: pkg/asyncdr/asyncdr.go [hunk 9]
```diff
@@ -97,6 +97,119 @@ func CreateMigration(
...
+				} else {
+					expected_migrations[mig.Name] = "Successful"
```
Comment (Rohit-PX) [inline pkg/asyncdr/asyncdr.go:182]:
```suggestion
					expectedMigrations[mig.Name] = "Successful"
```

File: pkg/asyncdr/asyncdr.go [hunk 10]
```diff
@@ -97,6 +97,119 @@ func CreateMigration(
...
+				}
+			}
+		}
+		log.InfoD("Waiting to complete %v migrations in %v time, %v completed as of now", count, migrationRetryTimeout, expected_migrations)
```
Comment (Rohit-PX) [inline pkg/asyncdr/asyncdr.go:186]:
```suggestion
		log.InfoD("Waiting to complete %v migrations in %v time, %v completed as of now", count, migrationRetryTimeout, expectedMigrations)
```

File: pkg/asyncdr/asyncdr.go [hunk 11]
```diff
@@ -97,6 +97,119 @@ func CreateMigration(
...
+		if len(expected_migrations) == count {
```
Comment (Rohit-PX) [inline pkg/asyncdr/asyncdr.go:187]:
```suggestion
		if len(expectedMigrations) == count {
```

File: pkg/asyncdr/asyncdr.go [hunk 12]
```diff
@@ -97,6 +97,119 @@ func CreateMigration(
...
+			return "", false, nil
+		}
+		return "", true, fmt.Errorf("some migrations are still pending")
+	}
+	_, err := task.DoRetryWithTimeout(checkNuOfMigrations, migTimeout, migInterval)
+	return expected_migrations, err
```
Comment (Rohit-PX) [inline pkg/asyncdr/asyncdr.go:193]:
```suggestion
	return expectedMigrations, err
```

File: tests/testTriggers.go [hunk 13]
```diff
@@ -6987,6 +7000,140 @@ func TriggerIopsBwAsyncDR(contexts *[]*scheduler.Context, recordChan *chan *Even
 	updateMetrics(*event)
 }
 
+func TriggerAsyncDRMigrationSchedule(contexts *[]*scheduler.Context, recordChan *chan *EventRecord) {
+	defer endLongevityTest()
+	startLongevityTest(AsyncDRMigrationSchedule)
+	defer ginkgo.GinkgoRecover()
+	log.InfoD("Async DR Migrationschedule triggered at: %v", time.Now())
```
Comment (Rohit-PX) [inline tests/testTriggers.go:7007]: It would be good to print the name of the migration schedule as well.

File: tests/testTriggers.go [hunk 14]
```diff
@@ -6987,6 +7000,140 @@ func TriggerIopsBwAsyncDR(contexts *[]*scheduler.Context, recordChan *chan *Even
...
+			currMigSched, createMigSchedErr := asyncdr.CreateMigrationSchedule(
+				migrationScheduleName, currMigNamespace, asyncdr.DefaultClusterPairName, currMigNamespace, &includeVolumesFlag,
+				&includeResourcesFlag, &startApplicationsFlag, schd_pol.Name, &suspendSched, autoSuspend)
+			if createMigSchedErr != nil {
+				UpdateOutcome(event, fmt.Errorf("failed to create migrationschedule wit error %v", err))
+				return
+			}
+			allMigrationsSched[currMigNamespace] = currMigSched.Name
+			time.Sleep(30 * time.Second)
+			mig_sched_resp, err := storkops.Instance().GetMigrationSchedule(currMigSched.Name, currMigNamespace)
```
Comment (Rohit-PX) [inline tests/testTriggers.go:7094]: Use camel case for all variables: `migSchedResp`

File: tests/testTriggers.go [hunk 15]
```diff
@@ -6987,6 +7000,140 @@ func TriggerIopsBwAsyncDR(contexts *[]*scheduler.Context, recordChan *chan *Even
...
+			if err != nil {
+				UpdateOutcome(event, fmt.Errorf("failed to get migrationschedule, error: %v", err))
+				return
+			}
+			if len(mig_sched_resp.Status.Items) == 0 {
```
Comment (Rohit-PX) [inline tests/testTriggers.go:7099]: Camel case

File: pkg/asyncdr/asyncdr.go (hunk 1 above)
Comment (gejain) [reply pkg/asyncdr/asyncdr.go]: Done!

Comment (gejain) [reply pkg/asyncdr/asyncdr.go]: Done!

File: pkg/asyncdr/asyncdr.go (hunk 2 above)
Comment (gejain) [reply pkg/asyncdr/asyncdr.go:110]: Added

File: pkg/asyncdr/asyncdr.go (hunk 3 above)
Comment (gejain) [reply pkg/asyncdr/asyncdr.go]: Added

Comment (gejain) [reply pkg/asyncdr/asyncdr.go]: Done!

File: pkg/asyncdr/asyncdr.go (hunk 4 above)
Comment (gejain) [reply pkg/asyncdr/asyncdr.go:184]: Agree!, I will add a config for implementing this based on app time.

File: pkg/asyncdr/asyncdr.go (hunk 5 above)
Comment (gejain) [reply pkg/asyncdr/asyncdr.go]: Done!

File: pkg/asyncdr/asyncdr.go (hunk 6 above)
Comment (gejain) [reply pkg/asyncdr/asyncdr.go]: Done!

File: pkg/asyncdr/asyncdr.go (hunk 8 above)
Comment (gejain) [reply pkg/asyncdr/asyncdr.go]: Done!

File: pkg/asyncdr/asyncdr.go (hunk 9 above)
Comment (gejain) [reply pkg/asyncdr/asyncdr.go]: Done!

File: pkg/asyncdr/asyncdr.go (hunk 10 above)
Comment (gejain) [reply pkg/asyncdr/asyncdr.go]: Done!

File: pkg/asyncdr/asyncdr.go (hunk 11 above)
Comment (gejain) [reply pkg/asyncdr/asyncdr.go]: Done!

File: pkg/asyncdr/asyncdr.go [hunk 16]
```diff
@@ -97,6 +97,119 @@ func CreateMigration(
...
+		}
+		migs := response.Status.Items["Interval"]
+		for _, mig := range migs {
+			var migrations []*storkapi.Migration
+			if _, ok := expected_migrations[mig.Name]; !ok {
+				migration, err := storkops.Instance().GetMigration(mig.Name, sched_namespace)
+				if err != nil {
+					return "", true, fmt.Errorf("failed to get migration: %v", mig.Name)
+				}
+				migrations = append(migrations, migration)
```
[the remark this answers is not in the document]
Comment (gejain) [reply pkg/asyncdr/asyncdr.go]: Implemented.

File: tests/testTriggers.go (hunk 13 above)
Comment (gejain) [reply tests/testTriggers.go:7007]: We cannot print here, as its just for trigger information.

File: tests/testTriggers.go (hunk 14 above)
Comment (gejain) [reply tests/testTriggers.go]: Done!

File: tests/testTriggers.go (hunk 15 above)
Comment (gejain) [reply tests/testTriggers.go]: Done!

**Review by Rohit-PX — APPROVED** [review]
Looks good to me.

**Comment by Rohit-PX** [comment]
@gejain : squash all commits to one and rebase with master branch, before merging.

**Pull request merged**